src/components/Editor/Board/Canvas/utils.ts

```diff
@@ -15,14 +15,108 @@ export function drawLine(context: CanvasRenderingContext2D, line: Line) {
   context.stroke();
 }
 
-export function createLine(pointY: number, pointX: number): Line {
+export function createLine(point: Point): Line {
   return {
     type: 'line',
-    points: [
-      {
-        x: pointX,
-        y: pointY,
-      },
-    ],
+    points: [point],
   } as Line;
 }
+
+// square distance between 2 points
+function getSqDist(p1: Point, p2: Point) {
+  const dx = p1.x - p2.x;
+  const dy = p1.y - p2.y;
+
+  return dx * dx + dy * dy;
+}
+
+// square distance from a point to a segment
+function getSqSegDist(p: Point, p1: Point, p2: Point) {
+  let { x, y } = p1;
+  let dx = p2.x - x;
+  let dy = p2.y - y;
+
+  if (dx !== 0 || dy !== 0) {
+    const t = ((p.x - x) * dx + (p.y - y) * dy) / (dx * dx + dy * dy);
+    if (t > 1) {
+      x = p2.x;
+      y = p2.y;
+    } else if (t > 0) {
+      x += dx * t;
+      y += dy * t;
+    }
+  }
+
+  dx = p.x - x;
+  dy = p.y - y;
+  return dx * dx + dy * dy;
+}
+
+function DouglasPeuckerStep(
+  points: Array<Point>,
+  first: number,
+  last: number,
+  tolerance: number,
+  simplified: Array<Point>,
+) {
+  let maxDist = tolerance;
+  let index = 0;
+
+  for (let i = first + 1; i < last; i += 1) {
+    const dist = getSqSegDist(points[i], points[first], points[last]);
+
+    if (dist > maxDist) {
+      index = i;
+      maxDist = dist;
+    }
+  }
+
+  if (maxDist > tolerance) {
+    if (first + 1 < index) {
+      DouglasPeuckerStep(points, first, index, tolerance, simplified);
+    }
+    simplified.push(points[index]);
+    if (index + 1 < last) {
+      DouglasPeuckerStep(points, index, last, tolerance, simplified);
+    }
+  }
+}
+
+function DouglasPeucker(points: Array<Point>, tolerance: number): Array<Point> {
```

Comment:
It would be nice to write the name of the function according to the convention.
https://github.com/basarat/typescript-book/blob/master/docs/styleguide/styleguide.md#variable-and-function

Reply:
Since it is a proper noun, I used uppercase letters, but I will unify them in lowercase letters